File: web/src/queries/cloudflareDns.ts
// GET/PUT/DELETE /api/v1/settings/cloudflare-dns
// (internal/api/cloudflare_dns.go's cloudflareDNSResource), mirroring
// queries/cloudflareTunnel.ts's shape for a distinct credential: a
// scoped Cloudflare API token (Zone:DNS:Edit) for ACME's DNS-01
// challenge, the only challenge type that can issue a wildcard
// certificate, not the cloudflared connector token.

import {
  queryOptions,
  useMutation,
  useQueryClient,
  useSuspenseQuery,
} from '@tanstack/react-query'
import { ApiError, readErrorMessage } from '../lib/apiError'

export const cloudflareDnsKeys = {
  all: ['cloudflare-dns'] as const,
}

// CloudflareDnsSettings mirrors cloudflareDNSResource exactly. The token
// itself never appears here: write-only on the request side, has_token
// reports presence instead.
export interface CloudflareDnsSettings {
  enabled: boolean
  has_token: boolean
}

export interface UpdateCloudflareDnsRequest {
  enabled: boolean
  // Empty/omitted means "leave the currently stored token unchanged".
  token?: string
}

export async function fetchCloudflareDnsSettings(): Promise<CloudflareDnsSettings> {
  const res = await fetch('/api/v1/settings/cloudflare-dns')
  if (!res.ok) {
    throw new ApiError(
      res.status,
      await readErrorMessage(
        res,
        `fetch cloudflare dns settings failed: ${res.status}`,
      ),
    )
  }
  return (await res.json()) as CloudflareDnsSettings
}

export function cloudflareDnsSettingsQueryOptions() {
  return queryOptions({
    queryKey: cloudflareDnsKeys.all,
    queryFn: fetchCloudflareDnsSettings,
    staleTime: 60_000,
  })
}

export function useCloudflareDnsSettings() {
  return useSuspenseQuery(cloudflareDnsSettingsQueryOptions())
}

// 501 means the control plane was started without APP_MASTER_KEY, the
// same server-configuration-gap case queries/cloudflareTunnel.ts's
// updateCloudflareTunnelSettings carries for the identical reason.
export async function updateCloudflareDnsSettings(
  req: UpdateCloudflareDnsRequest,
): Promise<CloudflareDnsSettings> {
  const res = await fetch('/api/v1/settings/cloudflare-dns', {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(req),
  })
  if (res.status === 501) {
    throw new ApiError(
      501,
      'Cloudflare DNS-01 requires a master key to be configured on this control plane.',
    )
  }
  if (!res.ok) {
    throw new ApiError(
      res.status,
      await readErrorMessage(
        res,
        `update cloudflare dns settings failed: ${res.status}`,
      ),
    )
  }
  return (await res.json()) as CloudflareDnsSettings
}

export function useUpdateCloudflareDnsSettings() {
  const queryClient = useQueryClient()
  return useMutation<
    CloudflareDnsSettings,
    ApiError,
    UpdateCloudflareDnsRequest
  >({
    mutationFn: updateCloudflareDnsSettings,
    onSuccess: (updated) => {
      queryClient.setQueryData(cloudflareDnsKeys.all, updated)
    },
  })
}

// DELETE /api/v1/settings/cloudflare-dns: disables DNS-01 and clears the
// stored token in one step. Returns the resulting resource, not 204, the
// same shape disconnectCloudflareTunnel establishes.
export async function disconnectCloudflareDns(): Promise<CloudflareDnsSettings> {
  const res = await fetch('/api/v1/settings/cloudflare-dns', {
    method: 'DELETE',
  })
  if (res.status === 501) {
    throw new ApiError(
      501,
      'Cloudflare DNS-01 requires a master key to be configured on this control plane.',
    )
  }
  if (!res.ok) {
    throw new ApiError(
      res.status,
      await readErrorMessage(
        res,
        `disconnect cloudflare dns failed: ${res.status}`,
      ),
    )
  }
  return (await res.json()) as CloudflareDnsSettings
}

export function useDisconnectCloudflareDns() {
  const queryClient = useQueryClient()
  return useMutation<CloudflareDnsSettings, ApiError, void>({
    mutationFn: disconnectCloudflareDns,
    onSuccess: (updated) => {
      queryClient.setQueryData(cloudflareDnsKeys.all, updated)
    },
  })
}

File: internal/api/cloudflare_dns.go
package api

import (
	"context"
	"encoding/json"
	"log/slog"
	"net/http"

	"github.com/GLINCKER/levelrail/internal/store"
)

// CloudflareDNSSecrets is the surface the Cloudflare DNS-01 settings
// handlers need from internal/secrets.Manager, the same shape
// CloudflareTunnelSecrets already establishes for a distinct credential
// (a scoped Cloudflare API token, not the cloudflared connector token).
type CloudflareDNSSecrets interface {
	SetValue(ctx context.Context, serviceName, envKey, plaintext string) error
	Exists(ctx context.Context, serviceName, envKey string) (bool, error)
	DeleteAll(ctx context.Context, serviceName string) error
}

// cloudflareDNSResource is the wire shape for GET, PUT, and DELETE
// /api/v1/settings/cloudflare-dns. The token itself never appears here
// in either direction, the same convention cloudflareTunnelResource
// establishes.
type cloudflareDNSResource struct {
	Enabled  bool `json:"enabled"`
	HasToken bool `json:"has_token"`
}

func (rt *Router) toCloudflareDNSResource(ctx context.Context, s store.CloudflareDNSSettings) cloudflareDNSResource {
	res := cloudflareDNSResource{Enabled: s.Enabled}
	if rt.cloudflareDNSSecrets != nil {
		exists, err := rt.cloudflareDNSSecrets.Exists(ctx, store.CloudflareDNSSecretsKey(), store.CloudflareDNSTokenEnvKey)
		if err != nil {
			rt.logger.Warn("api: check cloudflare dns token failed", slog.String("error", err.Error()))
		}
		res.HasToken = exists
	}
	return res
}

// handleGetCloudflareDNSSettings handles GET
// /api/v1/settings/cloudflare-dns. AbilityRead, matching GET
// /api/v1/settings/cloudflare-tunnel.
func (rt *Router) handleGetCloudflareDNSSettings(w http.ResponseWriter, r *http.Request) {
	settings, err := rt.cloudflareDNS.GetCloudflareDNSSettings(r.Context())
	if err != nil {
		rt.logger.Error("api: get cloudflare dns settings failed", slog.String("error", err.Error()))
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	writeJSON(w, http.StatusOK, rt.toCloudflareDNSResource(r.Context(), settings))
}

type updateCloudflareDNSRequest struct {
	Enabled bool `json:"enabled"`
	// Token is optional on update: empty means "leave the currently
	// stored token unchanged", the same convention
	// updateCloudflareTunnelRequest.Token already establishes.
	Token string `json:"token,omitempty"`
}

// handleUpdateCloudflareDNSSettings handles PUT
// /api/v1/settings/cloudflare-dns. AbilityRoot: instance-level
// infrastructure config, the same tier PUT /api/v1/settings/cloudflare-tunnel
// uses. Returns 501 without cloudflareDNSSecrets configured (no master
// key).
func (rt *Router) handleUpdateCloudflareDNSSettings(w http.ResponseWriter, r *http.Request) {
	if rt.cloudflareDNSSecrets == nil {
		writeError(w, http.StatusNotImplemented, "cloudflare dns is not configured on this control plane (no master key set)")
		return
	}

	var req updateCloudflareDNSRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	key := store.CloudflareDNSSecretsKey()
	hasToken := req.Token != ""
	if !hasToken {
		exists, err := rt.cloudflareDNSSecrets.Exists(r.Context(), key, store.CloudflareDNSTokenEnvKey)
		if err != nil {
			rt.logger.Error("api: check cloudflare dns token failed", slog.String("error", err.Error()))
			writeError(w, http.StatusInternalServerError, "internal error")
			return
		}
		hasToken = exists
	}
	if req.Enabled && !hasToken {
		writeError(w, http.StatusBadRequest, "token is required the first time cloudflare dns is enabled")
		return
	}

	if req.Token != "" {
		if err := rt.cloudflareDNSSecrets.SetValue(r.Context(), key, store.CloudflareDNSTokenEnvKey, req.Token); err != nil {
			rt.logger.Error("api: save cloudflare dns token failed", slog.String("error", err.Error()))
			writeError(w, http.StatusInternalServerError, "internal error")
			return
		}
	}

	settings := store.CloudflareDNSSettings{Enabled: req.Enabled}
	if err := rt.cloudflareDNS.UpdateCloudflareDNSSettings(r.Context(), settings); err != nil {
		rt.logger.Error("api: update cloudflare dns settings failed", slog.String("error", err.Error()))
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	writeJSON(w, http.StatusOK, rt.toCloudflareDNSResource(r.Context(), settings))
}

// handleDisconnectCloudflareDNS handles DELETE
// /api/v1/settings/cloudflare-dns: disables DNS-01 and clears the stored
// token. Idempotent, the same shape handleDisconnectCloudflareTunnel
// establishes.
func (rt *Router) handleDisconnectCloudflareDNS(w http.ResponseWriter, r *http.Request) {
	if rt.cloudflareDNSSecrets == nil {
		writeError(w, http.StatusNotImplemented, "cloudflare dns is not configured on this control plane (no master key set)")
		return
	}

	if err := rt.cloudflareDNSSecrets.DeleteAll(r.Context(), store.CloudflareDNSSecretsKey()); err != nil {
		rt.logger.Error("api: clear cloudflare dns token failed", slog.String("error", err.Error()))
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	settings := store.CloudflareDNSSettings{Enabled: false}
	if err := rt.cloudflareDNS.UpdateCloudflareDNSSettings(r.Context(), settings); err != nil {
		rt.logger.Error("api: disconnect cloudflare dns failed", slog.String("error", err.Error()))
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	writeJSON(w, http.StatusOK, rt.toCloudflareDNSResource(r.Context(), settings))
}
